Added Matlab implementation of generic SNS IK solvers for multiple tasks

robodreamer opened the pull request:
This commit adds the functions that implement SNS vel/acc IK algorithms
for multiple prioritized tasks. Testing codes have been updated as well
to validate the new functions.

Commit:
Added Matlab implementation of generic SNS IK solvers for multiple tasks

This commit adds the functions that implement SNS vel/acc IK algorithms
for multiple prioritized tasks. Testing codes have been updated as well
to validate the new functions.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_QP_mt.m:
```diff
@@ -0,0 +1,130 @@
+% INPUTS:
+%   ddqLow: lower limit for JDataoint accelerations
+%   ddqUpp: upper limit for JDataoint accelerations
+%   ddxGoalDataData: task-space accelerations of all the tasks in the order of priorities
```
Please add more data here. What is the data type of this input? How is it constructed and deconstructed?

robodreamer replied:
I did for all the functions.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_QP_mt.m:
```diff
@@ -0,0 +1,130 @@
+%   ddqLow: lower limit for JDataoint accelerations
+%   ddqUpp: upper limit for JDataoint accelerations
+%   ddxGoalDataData: task-space accelerations of all the tasks in the order of priorities
+%   JData: Jacobians of all the task in corresponding order
```
Same here: How are multiple jacobian matrices stored within a single input?

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_QP_mt.m:
```diff
@@ -0,0 +1,130 @@
+%   ddqUpp: upper limit for JDataoint accelerations
+%   ddxGoalDataData: task-space accelerations of all the tasks in the order of priorities
+%   JData: Jacobians of all the task in corresponding order
+%   dJdqData: dJ * dq of all the task in corresponding order
```
here too.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_QP_mt.m:
```diff
@@ -0,0 +1,130 @@
+problem.options = optimset('Display','off');
+problem.solver = 'quadprog';
+
+for iTask = 1:numTask
```
Is there a way to solve all tasks with a single QP solve? I see two possible reasons for this:
1) It might find a better solution than solving it sequentially
2) It might be faster to call the solver once.

Definitely keep this implementation - it is great. I'm more thinking as a possible additional version of the algorithm.

robodreamer replied:
I need to think a bit more about this. It is not straightforward since some of the tasks are infeasible, I don't know how the QP will figure the feasibility of each task when it is formulated in a big QP problem. I will leave this in TODO note and we will create another QP example when we figure it out.

MatthewPeterKelly replied:
That sounds good - thanks!

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_rr_mt.m:
```diff
@@ -0,0 +1,254 @@
+% INPUTS:
+%   ddqLow: lower limit for JDataoint accelerations
+%   ddqUpp: upper limit for JDataoint accelerations
+%   ddxGoalDataData: task-space accelerations of all the tasks in the order of priorities
```
Same comment here as in the previous file: make the data structure clear for each input.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_acc_rr_mt.m:
```diff
@@ -0,0 +1,254 @@
+
+%~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~%
+
+function taskScale = FindScaleFactor(low, upp, a)
```
It looks like this function is included in about eight different files. I propose that you put it in a stand-along file and then remove it from each of the solvers. Please also provide documentation for it.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_vel_QP_mt.m:
```diff
@@ -0,0 +1,125 @@
+%   dqLow: lower limit for joint velocities
+%   dqUpp: upper limit for joint velocities
+%   dxGoalData: task-space velocities of all the tasks in the order of priorities
+%   JData: Jacobians of all the task in the order of priorities
```
make clear that this and JData are cell arrays.

MatthewPeterKelly commented on sns_ik_lib/matlab/snsIk_vel_rr_mt.m:
```diff
@@ -0,0 +1,238 @@
+% INPUTS:
+%   dqLow: lower limit for joint velocities
+%   dqUpp: upper limit for joint velocities
+%   dxGoalData: task-space velocities of all the tasks in the order of priorities
```
cell array

MatthewPeterKelly reviewed (changes requested):
This looks great! 

I made a few minor suggestions, mostly relating to documentation.

Please move FindScaleFactor() into its own file and delete the duplicates.

Commit:
Addressed the reviewers' comments

- created a common function 'FindScaleFactor'
- clarified the input and output data types and sizes

robodreamer commented:
I have addressed the comments in the update.

MatthewPeterKelly reviewed (approved):
Looks great!
